src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PathSeqFilterSpark.java

```diff
@@ -1,268 +1,115 @@
-        final JavaRDD<GATKRead> readsBQFiltered = readsLengthFiltered.map(new ReadTransformerSparkifier(new BaseQualityReadTransformer(QUAL_PHRED_THRESH)));
-
-        //Filter reads with too many 'N's
-        final JavaRDD<GATKRead> readsAmbigFiltered = readsBQFiltered.filter(new ReadFilterSparkifier(new AmbiguousBaseReadFilter(FRAC_N_THRESHOLD)));
-        logger.info("Reads remaining after ambiguous base filtering: " + readsAmbigFiltered.count());
-
-        //Load Kmer hopscotch set and filter reads containing > 0 matching kmers
-        final JavaRDD<GATKRead> readsKmerFiltered = doKmerFiltering(ctx,readsAmbigFiltered);
-        logger.info("Reads remaining after kmer filtering: " + readsKmerFiltered.count());
-
-        //Filter unpaired reads
-        final JavaRDD<GATKRead> readsFilteredPaired = retainPairs(readsKmerFiltered);
-        logger.info("Reads remaining after unpaired filtering: " + readsFilteredPaired.count());
-
-        //BWA filtering against user-specified host organism reference
-        header = getHeaderForReads();
-        final JavaRDD<GATKRead> readsAligned = doHostBWA(ctx, header, readsFilteredPaired);
-
-        //Get unmapped reads (note these always come in pairs)
-        //TODO: retain read pairs by alignment score instead of flags
-        final JavaRDD<GATKRead> readsNonHost = readsAligned.filter(read -> read.isUnmapped() && read.mateIsUnmapped());
-        logger.info("Reads remaining after BWA filtering: " + readsFilteredPaired.count());
-
-        //TODO: repeat BWA with seed size 11
-
-        //Write output
-        header.setSortOrder(SAMFileHeader.SortOrder.queryname);
-        try {
-            ReadsSparkSink.writeReads(ctx, OUTPUT_PATH, null, readsNonHost, header, shardedOutput ? ReadsWriteFormat.SHARDED : ReadsWriteFormat.SINGLE);
-        } catch (final IOException e) {
-            throw new GATKException("Unable to write bam",e);
+        final SAMFileHeader header = getHeaderForReads();
+        if (filterArgs.alignedInput && (header.getSequenceDictionary() == null || header.getSequenceDictionary().isEmpty())) {
+            logger.warn("--isHostAligned is true but the BAM header contains no sequences");
+        }
+        if (!filterArgs.alignedInput && header.getSequenceDictionary() != null && !header.getSequenceDictionary().isEmpty()) {
+            logger.warn("--isHostAligned is false but there are one or more sequences in the BAM header");
+        }
+        header.setSequenceDictionary(new SAMSequenceDictionary());
+
+        try (final PSFilter filter = new PSFilter(ctx, filterArgs, getReads(), header)) {
+
+            final Tuple2<JavaRDD<GATKRead>, JavaRDD<GATKRead>> result = filter.doFilter();
+            final JavaRDD<GATKRead> pairedReads = result._1;
+            final JavaRDD<GATKRead> unpairedReads = result._2;
+
+            if (!pairedReads.isEmpty()) {
+                final SAMFileHeader.SortOrder oldSortOrder = header.getSortOrder();
```

Comment:
I thought you had already grouped the mate pairs together in the pairedReads list.  In which case, don't you want the writer to leave the order unperturbed, rather than doing what I fear is a global sort on query name?  And restoring the oldSortOrder at the end -- doesn't that leave you with a lie?  Could you just set the sort order as "unsorted" or "unknown" and leave it that way?